gpcontrib/arenadata_toolkit/src/arenadata_toolkit_guc.c

```diff
@@ -0,0 +1,248 @@
+#include "arenadata_toolkit_guc.h"
+
+#include "cdb/cdbvars.h"
+#include "catalog/objectaccess.h"
+#include "catalog/pg_db_role_setting.h"
+#include <limits.h>
+#include "utils/guc.h"
+#include "tf_shmem.h"
+
+int			bloom_size = DEFAULT_BLOOM_SIZE_BYTES;
+int			db_track_count = DEFAULT_DB_TRACK_COUNT;
+bool		is_tracked = DEFAULT_IS_TRACKED;
+bool		get_full_snapshot_on_recovery = DEFAULT_GET_FULL_SNAPSHOT_ON_RECOVERY;
+int			drops_count = DEFAULT_DROPS_COUNT;
+char	   *tracked_schemas = DEFAULT_TRACKED_SCHEMAS;
+char	   *tracked_rel_storages = DEFAULT_TRACKED_REL_STORAGES;
+char	   *tracked_rel_kinds = DEFAULT_TRACKED_REL_KINDS;
+int			tracking_worker_naptime_sec = DEFAULT_NAPTIME_SEC;
+
+/*
+ * Variable controlling GUC setting. Only extension functions are allowed
+ * to set GUC during NormalProcessing mode.
+ */
+static bool guc_is_unlocked = false;
+
+void
+tf_guc_unlock(void)
+{
+	guc_is_unlocked = true;
+}
+
+/*
+ * Prohibit changing the GUC value manually except several cases.
+ * This is not called for RESET, so RESET is not guarded
+ */
+static bool
+check_guc(GucSource source, bool *manual)
+{
+	if (IsInitProcessingMode() || Gp_role == GP_ROLE_EXECUTE ||
+		(Gp_role == GP_ROLE_DISPATCH && guc_is_unlocked))
+	{
+		guc_is_unlocked = false;
+
+		if (source != PGC_S_DATABASE &&
+			source != PGC_S_DEFAULT &&
+			source != PGC_S_TEST)
+			return false;
+
+		return true;
+	}
+
+	*manual = true;
+	return false;
+}
+
+/*
+ * Prohibit changing the arenadata_toolkit.tracking_is_db_tracked value manually
+ */
+static bool
+check_tracked(bool *newval, void **extra, GucSource source)
```

Comment:
It look that now the only difference among functions 
`check_tracked()`
`check_get_full_snapshot_on_recovery()`
`check_relkinds()`
`check_schemas()`
`check_relstorages()`
is the error message. 

What about separating the common part into a function? Thus the functions above can be reworked into a simplier implementation, like smth below:

```
static bool
check_relstorages(char **newval, void **extra, GucSource source)
{
	bool		manual = false;

	if (check_guc(source, &manual))
		return true;

	if (manual)
		GUC_check_errmsg("cannot change tracking status outside the tracking_register_relstorages function");
	return false;
}
```
->
```
static bool
check_relstorages(char **newval, void **extra, GucSource source)
{
	return some_internal_check(newval, extra, source, "cannot change tracking status outside the tracking_register_relstorages function");
}
```